### PR TITLE
Allow configuration of working directory

### PR DESCRIPTION
**Pull request by dfrankland**
Add a new`workingDirectory` property to the config to allow changing
the directory of new sessions.

### CODE REVIEW

File: config-default.js [hunk 1]
```diff
@@ -50,7 +50,11 @@ module.exports = {
 
     // the shell to run when spawning a new session (i.e. /usr/local/bin/fish)
     // if left empty, your system's login shell will be used by default
-    shell: ''
+    shell: '',
+
+    // the working directory to start in when spawning a new session (i.e. ~/projects)
```
Comment (rauchg) [inline config-default.js]: `e.g.` is correct here. If you can wrap `~/projects` in backticks it'd be nicer too ;)

File: config-default.js [hunk 2]
```diff
@@ -50,7 +50,11 @@ module.exports = {
...
+    // if left empty, your process.env.HOME directory will be used by default
```
Comment (rauchg) [inline config-default.js]: backticks around `process.env.HOME` would be nice too!

File: index.js [hunk 3]
```diff
@@ -4,6 +4,7 @@ const createMenu = require('./menu');
 const Session = require('./session');
 const genUid = require('uid2');
 const { resolve } = require('path');
+const untildify = require('untildify');
```
Comment (rauchg) [inline index.js]: amazing. @sindresorhus has created every single module

**Comment by rauchg** [comment]
+1 other than the minor nits

File: index.js (hunk 3 above)
Comment (dfrankland) [inline index.js]: What would we do without you @sindresorhus 🎱

**Comment by dfrankland** [comment]
👍 I have no problems with nitpicks. I have made the changes requested.

**Comment by developit** [comment]
@rauchg will this break [hypercwd](https://github.com/hharnisc/hypercwd)? Or will it just provide a default that is then overridden by it? (thinking it's actually the latter since hypercwd would be providing the `cwd` option via rpc)

**Comment by rauchg** [comment]
Default option. Pinging @hharnisc so that he can review this.

**Comment by developit** [comment]
Awesome - I've been getting some cases where new tabs open to `/`, hopefully this fixes.

**Comment by hharnisc** [comment]
@rauchg I'll take a look at this in detail this evening. Thanks for the heads
up!  
  
\- Harrison  


On Jul 18 2016, at 11:53 am, Jason Miller &lt;notifications@github.com&gt;
wrote:  

> Awesome - I've been getting some cases where new tabs open to /, hopefully
this fixes.

>

> —  
> You are receiving this because you were mentioned.  
> Reply to this email directly, [view it on
GitHub](https://github.com/zeit/hyperterm/pull/219#issuecomment-233388779&r=cmVwbHkrMDAxNTJlMmZmN2FkYWU3ZDAzNzVmMWZkMjFjNzRkODliYzU3ZThkM2Q2ZWM0ZTc1OTJjZjAwMDAwMDAxMTNhNGNiYTE5MmExNjljZTA5ZTUwYTJjQHJlcGx5LmdpdGh1Yi5jb20=),
or [mute the thread](https://github.com/notifications/unsubscribe-auth
/ABUuL9PDCaUkBrMrcfb_b_dVKefMf-
aMks5qW6-hgaJpZM4JOWpQ&r=cmVwbHkrMDAxNTJlMmZmN2FkYWU3ZDAzNzVmMWZkMjFjNzRkODliYzU3ZThkM2Q2ZWM0ZTc1OTJjZjAwMDAwMDAxMTNhNGNiYTE5MmExNjljZTA5ZTUwYTJjQHJlcGx5LmdpdGh1Yi5jb20=).

**Comment by hharnisc** [comment]
Giving this a spin now, here goes nothing!

**Comment by hharnisc** [comment]
So this works and doesn't interfere with the `hypercwd` plugin. 👍 

It is pretty nice to be able to specify where the first terminal tab opens. Nice one @dfrankland!

I have to ask before making the call to merge this though @rauchg. Could this be a plugin instead?

If we've got a common place to store plugin specific config parameters this seems like a nice simple plugin to implement. 

No matter what though I would really like to see the cwd pass through `untildify`

```js
rpc.on('new', ({ rows = 40, cols = 100, cwd }) => {
      cwd = cwd ? untildify(cwd) : process.env.HOME;
```

Plugin would look something like this

```js
const getCwd = () => {
  // get the workingDirectory param from the config somehow
};

exports.middleware = (store) => (next) => (action) => {
  switch (action.type) {
    case 'INIT':
      store.dispatch({
        type: 'UPDATE_CWD',
        cwd: getCwd()
      });
      break;
  }
  next(action);
};

exports.reduceUI = (state, action) => {
  switch (action.type) {
    case 'UPDATE_CWD':
      return state.set('cwd', action.cwd);
  }
  return state;
};
```

**Comment by hharnisc** [comment]
So if #271 gets accepted you'll only need the following for a plugin 

```js
const getCwd = () => {
  // get the workingDirectory param from the config somehow
};

exports.middleware = (store) => (next) => (action) => {
  switch (action.type) {
    case 'INIT':
      store.dispatch({
        type: 'SESSION_SET_CWD',
        cwd: getCwd()
      });
      break;
  }
  next(action);
};
```

**Comment by rauchg** [comment]
@hharnisc I like that idea. What if instead you add the option for a default path to `hypercwd`? I kinda like the idea of `hypercwd` taking care of all the working directory needs :)

**Comment by rauchg** [comment]
Thinking more about this, I think having that in `hypercwd` or even another plugin is a good idea. It's also a good example to have in `-website` about how plugins can read custom configuration flags from `.hyperterm.js` by the way.

**Comment by bevacqua** [comment]
See [`hyperterm-working-directory`](https://github.com/bevacqua/hyperterm-working-directory) :tada:

**Pull request closed without merging**